### ISSUE TITLE
Ability to assign scale and rotation to prefabs

### ISSUE THREAD

**Issue by LaraEngebretson** [AtomicGameEngine/AtomicGameEngine#328]

**Comment by Alan-FGR** [comment]
@mattbenic are we going to revamp the prefab system? Can we just take advantage of built-in xml serialization or is it going to be hard to make it work with script components?

**Comment by mattbenic** [comment]
So there's a much bigger issue here, one that's been a major pain for ages: objects are either 100% linked to prefabs, or 100% disconnected. It's not possible to have an object based on a prefab but with one or two values changed on just that object (like in Unity).
This missing functionality actually makes prefabs much less useful, since the use cases for identical objects in a scene with zero changes between them isn't nearly as big.

I don't know enough about the way the prefab system is implemented to say if this is a rewrite, but it's definitely an enhancement we'd want at some point.

**Comment by Alan-FGR** [comment]
This pretty much only applies to the editor/scene: when you're instantiating a prefab from code it's inevitable that it will have the saved values upon creation that you later change, we could have a special overload to override the positional data right away though, what could be specially important if you have physics components in it so you instantiate it at the correct position already instead of moving it, what requires reinsertion in the spatial optimization systems both in the built-in scenegraph and physics engine.

With that in mind I think the best way to solve this is in the Scene data, maybe in each scene we could store in a dictionary the non-default values for the prefabs, so when the scene is created it applies them possibly via reflection. We could even store this data separately actually, we just need a dictionary of prefabs nodes IDs, in which values are dictionaries of component reflection data ("ComponentType+MemberName") and values. May get tricky when such values are complex data types.

**Comment by mattbenic** [comment]
Looking at this again, I think what Lara described is exactly the opposite, being able to have prefabs have location data so when you place them they're positioned in the scene. TBH that makes no sense to me.

On overriding values in the scene, yeah we could basically just serialize just those parts of the object that are different alongside the reference to the prefab. When deserializing we apply those over an already instantiated copy of the prefab. But yeah this may actually be a completely different issue which I'm fairly sure exists elsewhere.

**Comment by Alan-FGR** [comment]
Well, I think it's about the ability to assign rotation and scale to prefabs on the scene that are different than the values serialized in the prefab itself 😕

**Comment by Alan-FGR** [comment]
Btw, in Unity prefabs do have positional data serialized, so if you instantiate them without passing any of that data they use the default values, say for example it's part of a scene you're loading at runtime from prefabs, etc.

**Comment by mattbenic** [comment]
We can already do that though (position and scale prefab instances in the scene)

**Comment by Alan-FGR** [comment]
Yep, prefabbed objects need to load the values from the prefabs except when they are overriden in the scene. I'm going to take a look on how Duality solved that, they have a system like Unity and their code is really great albeit a bit overengineered to my taste.
